Entity/Components/Stats/EntityStats.gd

```diff
@@ -83,3 +83,10 @@ static func get_value_modified_by_stats(modifier_name: GlobalVar.POSSIBLE_MODIFI
 		modified_value = target_stats._calculate_modified_value_defense(modifier_name, modified_value) 
 
 	return modified_value
+
+
+# This function is only used to setup the dictionary in tests
+# In play, the tree logic will already properly setup everything
+func ready_entity_stats() -> void:
+	_offense_modifier_dict.ready_stat_dict()
+	_defense_modifier_dict.ready_stat_dict()
```

Comment:
Took me a minute to wrap my head around the reasoning for having separate ready functions for test cases. If my understanding is correct, those are needed because _init is called before `@export` variables are updated? So when we are initializing things through the scene we have to do a call_deferred to make sure the export variables are initialized. But when we are doing our test cases we don't care about the export variables, so we call some custom ready functions to make sure things are set up properly.

An alternative I thought of was to add a `on_stat_modifier_initialized` signal to `StatModifiers`. Then put that signal in the `ready` of `StatModifiers` like so:
```GDScript
func ready(	_permanent_add: int = permanent_add, 
			_permanent_multiply: float = permanent_multiply, 
			_temporary_add: int = temporary_add, 
			_temporary_multiply: float = temporary_multiply) -> void:
	modifier_base_dict = {   
					_MODIFIER_KEYS.PERMANENT_ADD:        _permanent_add,  # int
					_MODIFIER_KEYS.PERMANENT_MULTIPLY:   _permanent_multiply,  # float
					_MODIFIER_KEYS.TEMPORARY_ADD:        _temporary_add,  # int
					_MODIFIER_KEYS.TEMPORARY_MULTIPLY:   _temporary_multiply   # float
				}
	on_stat_modifier_initialized.emit()
```
Then only start modifying stats in test cases after we receive the signal. Like so:
```GDScript
# apply strength to player and damage enemy
func test_strength_status():
	var strength_status: Buff_Strength = Buff_Strength.new()
	# add 1 damage offense on the player
	var stat_modifer: StatModifiers = StatModifiers.new()
	stat_modifer.temporary_add = 1

	await stat_modifer.on_stat_modifier_initialized

	strength_status.status_modifier_base_value = stat_modifer
	_player.get_status_component().add_status(strength_status, _player)
```
We will probably need a similar signal when testing Buffs/Debuffs and cards with status, because those also use the stat modifications. Any thoughts?

Reply:
The reason for the call deferred is because as you guessed, the export variables are not taken into account, because i have to create a new object in the editor to be able to modify the `@export` variables, but by doing so i'm already calling `_init` which leads to the modification of the `@export` variables not being taken into account.
With the `call_deferred`, when playing normally and using the scene tree, the `_init` call is done after the export of variables which allows for editing through the inspector (which is what we want).

I don't think the solution you are giving would directly work:
```GDScript
func test_strength_status():
	var strength_status: Buff_Strength = Buff_Strength.new()
	# the buff gets created, no problem here
	# add 1 damage offense on the player
	var stat_modifer: StatModifiers = StatModifiers.new()
	# the stat_modifier is created, but because of the call_deferred ready is not called (since we are not using the tree in the tests
	stat_modifer.temporary_add = 1
	# you modify the export variable, yes you can do that, no problem here

	await stat_modifer.on_stat_modifier_initialized
	# the ready function was never called (I'm not sure the _init would even if you wait, so you will just be stuck waiting for the signal

	strength_status.status_modifier_base_value = stat_modifer
	_player.get_status_component().add_status(strength_status, _player)
```

I'll quickly put that in my current code to see if it actually works or not

Reply:
Alright nevermind, that seems to work with a signal. i'm not sure what is the real advantage over calling the `ready` function of `StatModifiers` directly though. Because yes I'm awaiting the signal for this specific case, but when i want to ready the entire offense/defense dictionaries, if I simply `await` for the signal i would continue as soon as I receive one such signal, ie as soon as at least one stat_modifier is ready. And I can't think of a good way to tell it to wait for **all** stat_modifier to be ready. Calling the ready function seems more reliable.

You also say "We will probably need a similar signal when testing Buffs/Debuffs and cards with status, because those also use the stat modifications" but that's already what we do.
We have status like strength, weakness, vulnerability. We even have a status that buffs the time of the poison being applied `Buff_Poison_Duration`

Reply:
I was thinking the signal might be better since it removes the need to have unique code that will only be used for test cases. I worry a bit about that, because its functioning slightly different from the actual game. But the case you mentioned with the StatDictBase is definitely tricky. You'd probably have to add a new signal in StatDictBase that fires when all the StatModifiers are initialized.

Reply:
If you think about it, the signals also will be used only by the tests in our case; it's also code relevant only to tests because the actual game doesn't await the signals.

If we want to use signals, I think a solution is so add a counter set to 0 at the `StatDictBase` level as well as a signal `on_stat_dict_base_ready`. The counter gets one more each time there is a signal received from `StatModifiers`, when it reaches the size of `Global_var.MODIFIER_KEYS` then it emits
Now the question I have is, what happens with several dictionaries / enemy. The counter is for each dictionary, so there won't be a problem of shared value that will only trigger the signal once, but if we await this signal how do we know that each offense / defense dictionary of each entity is properly setup ?

It would mean I also have to add a signal at the `EntityStats` level with a counter going to 2 (offense and defense) and then emit to say this entity stats are ready. For the entities it's probably fine since we setup sequentially. 
But that's 3 signals and 2 counters to add for each entity basically. For something that would only be useful to tests. I think from a debug standpoint, the functions I currently have are probably easier to debug since it's just recursive calls and for loop.

Reply:
Yeah that's fair. Let's keep your solution then